bundledatasrc_test.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package charm
+
+import (
+	"archive/zip"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/juju/testing"
+	gc "gopkg.in/check.v1"
+)
+
+type BundleDataSourceSuite struct {
+	testing.IsolationSuite
+}
+
+var (
+	_ = gc.Suite(&BundleDataSourceSuite{})
+
+	multidocBundle = `
+applications:
+  wordpress:
+    charm: wordpress
+  mysql:
+    charm: mysql
+    num_units: 1
+relations:
+  - ["wordpress:db", "mysql:server"]
+--- # overlay.yaml
+applications:
+  wordpress:
+    offers:
+      offer1:
+        endpoints:
+          - "some-endpoint"
+--- # overlay2.yaml
+applications:
+  wordpress:
+    offers:
+      offer1:
+        acl:
+          admin: "admin"
+          foo: "consume"
+`
+)
+
+func (s *BundleDataSourceSuite) TestReadMultiDocBundle(c *gc.C) {
+	specs := []struct {
+		descr     string
+		getSource func() (BundleDataSource, error)
+	}{
+		{
+			descr: "reading bundle from local yaml file",
+			getSource: func() (BundleDataSource, error) {
+				path := bundleDirPath(c, "wordpress-multidoc")
+				return LocalBundleDataSource(filepath.Join(path, "bundle.yaml"))
+			},
+		},
+		{
+			descr: "reading bundle from exploded archive folder",
+			getSource: func() (BundleDataSource, error) {
+				path := bundleDirPath(c, "wordpress-multidoc")
+				return LocalBundleDataSource(path)
+			},
+		},
+		{
+			descr: "reading bundle from archive",
+			getSource: func() (BundleDataSource, error) {
+				path := archiveBundleDirPath(c, "wordpress-multidoc")
+				return LocalBundleDataSource(path)
+			},
+		},
+		{
+			descr: "reading bundle from stream",
+			getSource: func() (BundleDataSource, error) {
+				r := strings.NewReader(multidocBundle)
+				return StreamBundleDataSource(r, "https://example.com")
+			},
+		},
+	}
+
+	for specIndex, spec := range specs {
+		c.Logf("[test %d] %s", specIndex, spec.descr)
+
+		ds, err := spec.getSource()
+		c.Assert(err, gc.IsNil)
+
+		parts := ds.Parts()
+		c.Assert(parts, gc.HasLen, 3)
+		s.assertFieldPresent(c, parts[1], "applications.wordpress.offers.offer1.endpoints")
+		s.assertFieldPresent(c, parts[2], "applications.wordpress.offers.offer1.acl.admin")
+	}
+}
+
+func (s *BundleDataSourceSuite) TestResolveAbsoluteFileInclude(c *gc.C) {
+	target, err := filepath.Abs(filepath.Join(c.MkDir(), "example"))
+	c.Assert(err, gc.IsNil)
+
+	expContent := []byte("example content")
+	c.Assert(ioutil.WriteFile(target, expContent, os.ModePerm), gc.IsNil)
+
+	ds := new(resolvedBundleDataSource)
+
+	got, err := ds.ResolveInclude(target)
+	c.Assert(err, gc.IsNil)
+	c.Assert(got, gc.DeepEquals, expContent)
```

Comment:
we probably also would rather compare (string(got)) because otherwise it tells you [0x32, 0x23,} != [0x12, 0x45]

which is usually not very nice to understand as a human. :)